### PR TITLE
Retryer: replace an instance of Date with an epoch millisecond

### LINKED ISSUE #2154: ErrorDecoder and retry-after header timezones issue
I believe there is a timezone issue in the the core/src/main/java/feign/codec/ErrorDecoder.java when parsing the retry-after header in the "public Date apply(String retryAfter)" method.

The code in the mentioned class basically does:

```
public static void main(String[] args) throws ParseException {
    final DateFormat RFC822_FORMAT = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss 'GMT'", Locale.US);
    Date d = RFC822_FORMAT.parse(RETRY-HEADER);
}
```

I'm in the timezone Europe/Berlin, here a "Mon, 14 Aug 2023 13:32:06 GMT" retry-after header becomes a "Mon Aug 14 13:32:06 CEST 2023" Date object .

Unfortunately the RetryableException that is then thrown and handled by my customized Retryer just contains the Date object and not the raw response header, so I cannot fix it by parsing the retry-after header again.

I created a local workaround with a custom errordecoder (the convert functrion is the workaround itself, going through ZonedDateTime and invoking Instant):

```
public class CustomErrorDecoder implements ErrorDecoder {
    @Override
    public Exception decode(String methodKey, Response response) {
        FeignException exception = feign.FeignException.errorStatus(methodKey, response);
        String retryAfter = firstOrNull(response.headers(), RETRY_AFTER);
        if (retryAfter != null) {
            return new RetryableException(
                    response.status(),
                    exception.getMessage(),
                    response.request().httpMethod(),
                    exception,
                    convert(retryAfter),
                    response.request());
        }
        return exception;
    }

    private Date convert(String retryAfter) {
        try {
            ZonedDateTime zoned = ZonedDateTime.parse(retryAfter, DateTimeFormatter.RFC_1123_DATE_TIME);
            return Date.from(zoned.toInstant());
        } catch(Exception e) {

        }

        return null;
    }

    private <T> T firstOrNull(Map<String, Collection<T>> map, String key) {
        if (map.containsKey(key) && !map.get(key).isEmpty()) {
            return map.get(key).iterator().next();
        }
        return null;
    }
}
```

### PR DESCRIPTION
**Pull request by vitalijr2**
Resolves #2154 

We do not use Date as date, every time we calculate epoch milliseconds. The Date class has issues with timezone.

### Summary by CodeRabbit

- Refactor: Replaced usage of `Date` with `Clock` in various classes and methods for better time handling.
- New Feature: Introduced a custom `TestClock` class for more flexible testing.
- Refactor: Updated `RetryableException` to use `Long` instead of `Date` for retry-after value, providing more flexibility.
- Refactor: Improved retry mechanism in `Retryer` using exponential backoff.
- Refactor: Replaced usage of `DateFormat` with `DateTimeFormatter` in `RetryAfterDecoder` for modern date/time handling.
- Test: Updated test cases to reflect changes in the main codebase.
- Chore: Updated build configuration to exclude certain file types from the build process.

### CODE REVIEW

File: core/src/main/java/feign/FeignException.java [hunk 1]
```diff
@@ -276,7 +276,7 @@ static FeignException errorExecuting(Request request, IOException cause) {
         format("%s executing %s %s", cause.getMessage(), request.httpMethod(), request.url()),
         request.httpMethod(),
         cause,
-        null, request);
+        (Long) null, request);
```
[the remark this answers is not in the document]
Comment (vitalijr2) [reply core/src/main/java/feign/FeignException.java:279]: It produces `error: reference to RetryableException is ambiguous` error.

File: core/src/main/java/feign/RetryableException.java [hunk 2]
```diff
@@ -24,14 +24,22 @@
...
+      Long retryAfter, Request request) {
+    super(status, message, request, cause);
+    this.httpMethod = httpMethod;
+    this.retryAfter = retryAfter;
+  }
+
+  @Deprecated
   public RetryableException(int status, String message, HttpMethod httpMethod, Throwable cause,
       Date retryAfter, Request request) {
     super(status, message, request, cause);
```
[the remark this answers is not in the document]
Comment (vitalijr2) [reply core/src/main/java/feign/RetryableException.java:45]: I think at least @velo and @kdavisk6 should approve.

File: core/src/main/java/feign/FeignException.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vitalijr2) [reply core/src/main/java/feign/FeignException.java]: It was implemented, see new comment here https://github.com/OpenFeign/feign/pull/2170#discussion_r1325785615

File: core/src/main/java/feign/codec/ErrorDecoder.java [hunk 3]
```diff
@@ -16,16 +16,15 @@
...
-import static java.util.Locale.US;
+import static java.time.format.DateTimeFormatter.RFC_1123_DATE_TIME;
 import static java.util.concurrent.TimeUnit.SECONDS;
 import feign.FeignException;
 import feign.Response;
 import feign.RetryableException;
-import java.text.DateFormat;
-import java.text.ParseException;
-import java.text.SimpleDateFormat;
+import java.time.ZonedDateTime;
```
Comment (velo) [inline core/src/main/java/feign/codec/ErrorDecoder.java:24]: It's this available on java8? Nice

**Review by velo — APPROVED** [review]

**Pull request merged**